## **Function**

Question 1: A company stores monthly revenue data that should not be modified accidentally.
1. Would you use a list or tuple? Justify.
2. Create a tuple of quarterly revenues.
3. Try modifying one value. What happens? Explain why.
**bold text**

#Answer:
1. Would you use a list or tuple? Justify.
   .I would use a tuple because monthly revenue data should not be modified accidentally. Tuples are immutable, which means their values cannot be changed after creation. This helps protect important financial data from unintended changes.


In [1]:
#2.Create a tuple of quarterly revenues.
quarterly_revenues = (250000, 300000, 275000, 325000)
print(quarterly_revenues)


(250000, 300000, 275000, 325000)


In [3]:
#3. Try modifying one value. What happens? Explain why.
quarterly_revenues[1] = 350000

TypeError: 'tuple' object does not support item assignment

Explanation:
This error occurs because a tuple is immutable. Once a tuple is created, its elements cannot be changed, added, or removed. This makes tuples suitable for storing important data, such as company revenue records, that should remain unchanged

#Question 2: You are given student performance data:
students = {
"Amit": 85,
"Sara": 92,
"John": 78,
"Meera": 88
}
Tasks:
1. Add a new student.
2. Update marks of one student.
3. Find the student with the highest marks.
4. Convert dictionary keys into a list.
5. Explain why dictionaries are important in Data Science workflows.

Answer 2:
1. Add a new student

In [5]:
students = { "Amit": 85, "Sara": 92,"John": 78,"Meera": 88 }
students["Riya"] = 90
print(students)

{'Amit': 85, 'Sara': 92, 'John': 78, 'Meera': 88, 'Riya': 90}


2. Update marks of one student

In [6]:
students = { "Amit": 85, "Sara": 92,"John": 78,"Meera": 88, "Riya": 90 }
students["John"] = 82
print(students)


{'Amit': 85, 'Sara': 92, 'John': 82, 'Meera': 88, 'Riya': 90}


3. Find the student with the highest marks

In [7]:
students = { "Amit": 85, "Sara": 92,"John": 78,"Meera": 88, "Riya": 90 }
highest_marks = max(students, key=students.get)
print(highest_marks)

Sara


4. Convert dictionary keys into a list

In [8]:
students = { "Amit": 85, "Sara": 92,"John": 78,"Meera": 88, "Riya": 90 }
student_list = list(students.keys())
print(student_list)

['Amit', 'Sara', 'John', 'Meera', 'Riya']


5. Explain why dictionaries are important in Data Science workflows

Dictionaries are important in Data Science because:
.They store data in key-value pairs, making information easy to organize.
.They provide fast access to data using keys.
.They are useful for storing feature values, model parameters, and results.
.They help represent JSON data, which is commonly used in APIs and data exchange.
.They make data processing and analysis more efficient by allowing quick lookups and updates
Example: A dictionary can store student names as keys and their marks as values, making it easy to retrieve and update information

#Question 3:  An e-commerce platform has customer IDs from two campaigns:
campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}
1. Find unique customers across both campaigns.
2. Find customers common in both.
3. Find customers exclusive to campaign1.
4. Explain how sets help in removing duplicates in real DS projects.

Answer 3:

Given sets:

In [ ]:
campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}

1. Find unique customers across both campaigns

    Use the union (|) operation:

In [10]:
campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}
unique_customers = campaign1 | campaign2
print(unique_customers)

{101, 102, 103, 104, 105, 106, 107}


2. Find customers common in both campaigns

    Use the intersection (&) operation:

In [11]:
campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}
common_customers = campaign1 & campaign2
print(common_customers)

{104, 105}


3. Find customers exclusive to campaign1

    Use the difference (-) operation:

In [12]:
campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}
exclusive_campaign1 = campaign1 - campaign2
print(exclusive_campaign1)

{101, 102, 103}


4. Explain how sets help in removing duplicates in real Data Science projects

    Sets are very useful in Data Science because they automatically store only unique values.

Benefits:


.Remove duplicate records from datasets.
.Identify unique customers, products, or transactions.
.Improve data quality before analysis.
.Perform fast operations such as union, intersection, and difference.
.Reduce storage and processing time when working with repeated data.

In [13]:
#Example

campaign1 = {101, 102, 103, 104, 105}
campaign2 = {104, 105, 106, 107}
customer_ids = [101, 102, 102, 103, 104, 104, 105]

unique_ids = set(customer_ids)
print(unique_ids)

{101, 102, 103, 104, 105}


Question 4: You receive raw CSV data where:
price = "499.99"
quantity = "5"
1. Convert them to correct numeric types.
2. Compute total revenue.
3. Explain why typecasting errors can break ML pipelines.

Answer4:
  Given Data:

In [ ]:
price = "499.99"
quantity = "5"

1. Convert them to correct numeric types

In [14]:
price = "499.99"
quantity = "5"
price = float(price)      # Convert string to float
quantity = int(quantity)  # Convert string to integer

print(price)
print(quantity)

499.99
5


2. Compute total revenue

In [16]:
price = "499.99"
quantity = "5"
price = float(price)      # Convert string to float
quantity = int(quantity)  # Convert string to integer
total_revenue = price * quantity
print(total_revenue)



2499.95


Calculation:

499.99×5=2499.95

So, the total revenue = 2499.95.

3. Explain why typecasting errors can break ML pipelines

Typecasting is the process of converting data from one type to another (e.g., string to integer or float).

Why typecasting errors are problematic in Machine Learning:

1.Mathematical operations fail


 .ML algorithms require numeric data.                                     
 .If numbers are stored as strings, calculations cannot be performed correctly.

In [19]:
"5" * "10"   # Error

TypeError: can't multiply sequence by non-int of type 'str'

2.Incorrect results

.String operations behave differently from numeric operations.

In [17]:
"5" + "10"   # Output: "510"
5 + 10       # Output: 15

15

3.Model training errors

.Most ML libraries (such as NumPy, Pandas, and Scikit-learn) expect numerical features.

.String values may cause the model to fail during training.

4.Data quality issues

.Incorrect data types can lead to wrong predictions and unreliable analysis.

5.Pipeline failures

.A single type mismatch can stop an entire ETL (Extract, Transform, Load) or ML workflow.

Example:


In [20]:
price = "499.99"
quantity = "5"

# Without typecasting
total = price * quantity   # Error

TypeError: can't multiply sequence by non-int of type 'str'

After proper typecasting:

In [21]:
price = "499.99"
quantity = "5"
total = float(price) * int(quantity)
print(total)

2499.95


Conclusion:

Typecasting ensures data is in the correct format for calculations, analysis, and machine learning models. Incorrect data types can cause errors, inaccurate results, and complete pipeline failures.

Question 5:

 Given: sales = [12000, 15000, 11000, 18000, 17000] Using a loop:
  1. Calculate total revenue.
  2. Count how many months had sales above 15000.
  3. Find maximum sales without using built-in max()

Answer5:

In [22]:
sales = [12000, 15000, 11000, 18000, 17000]

total_revenue = 0
months_above_15000 = 0
max_sales = sales[0]  # Start with first element as initial maximum

for sale in sales:
    # 1. Add each sale to total
    total_revenue += sale

    # 2. Check if sale is strictly above 15000
    if sale > 15000:
        months_above_15000 += 1

    # 3. Update max if current sale is greater
    if sale > max_sales:
        max_sales = sale

print(f"Total Revenue: ₹{total_revenue}")
print(f"Months above ₹15,000: {months_above_15000}")
print(f"Maximum Sales: ₹{max_sales}")

Total Revenue: ₹73000
Months above ₹15,000: 2
Maximum Sales: ₹18000


Answer6:

Given:ratings = [
    [5, 4, 3],
    [4, 4, 5],
    [3, 5, 4]
]

1. Compute Average Rating

In [24]:
ratings =[ [5, 4, 3],[4, 4, 5],[3, 5, 4]]

total = 0
count = 0

for row in ratings:
    for rating in row:
        total += rating
        count += 1

average = total / count
print("Average Rating:", average)


Average Rating: 4.111111111111111


2. Print All Ratings Greater Than 4

In [25]:
ratings = [
    [5, 4, 3],
    [4, 4, 5],
    [3, 5, 4]
]
for row in ratings:
    for rating in row:
        if rating > 4:
            print(rating)


5
5
5


3. How Nested Loops Are Used in Matrix-Based ML Computations

   A matrix is a 2D structure (rows and columns). Nested loops are used to access every element in the matrix.

In [27]:
ratings = [
    [5, 4, 3],
    [4, 4, 5],
    [3, 5, 4]
]

for row in matrix:
    for value in row:

SyntaxError: incomplete input (991657775.py, line 8)

In Machine Learning, nested loops are commonly used for:

 .Matrix multiplication (combining input data with weights)

 .Computing distances between data points

 .Updating weights during training

 .Image processing, where each pixel is treated as a matrix element

 .Calculating loss and gradients across datasets

 Example:

In [30]:
ratings = [
    [5, 4, 3],
    [4, 4, 5],
    [3, 5, 4]
]
for i in range(rows):
    for j in range(cols):
        result[i][j] = matrix1[i][j] + matrix2[i][j]

Here, the outer loop iterates through rows and the inner loop iterates through columns, allowing operations on every matrix element. This element-by-element processing is fundamental to many ML algorithms.

Question 7:

   Write a function:
def calculate_revenue(price, quantity):
that returns total revenue.
Then:
1. Call it for 3 different products.
2. Explain why modular functions are important in large DS projects.

Answer7:

1.Call it for 3 different products.

In [31]:
# Function to calculate revenue
def calculate_revenue(price, quantity):
    return price * quantity

# 1. Call it for 3 different products
product1_revenue = calculate_revenue(50, 100)   # Product A
product2_revenue = calculate_revenue(120, 40)   # Product B
product3_revenue = calculate_revenue(75, 60)    # Product C

print("Product A Revenue:", product1_revenue)
print("Product B Revenue:", product2_revenue)
print("Product C Revenue:", product3_revenue)

Product A Revenue: 5000
Product B Revenue: 4800
Product C Revenue: 4500


2. Why are modular functions important in      large Data Science (DS) projects?

1.Code Reusability – A function can be used multiple times without rewriting the same code.

2.Easy Maintenance – Changes only need to be made in one place, reducing errors.

3.Better Readability – Breaking a project into smaller functions makes the code easier to understand.

4.Simpler Testing and Debugging – Each function can be tested independently to find issues quickly.

5.Team Collaboration – Different team members can work on separate functions or modules without interfering with each other.

In large DS projects, modular functions help organize complex workflows such as data cleaning, feature engineering, model training, and evaluation, making the project more scalable and manageable.

Question 8:

 Given
prices = [500, 1500, 2500, 800, 1200]
1. Use map() + lambda to add 18% GST to all prices.
2. Use filter() to select prices above 1000.
3. Explain where such transformations are used in real-world pipelines

Answer8:

Given:
prices = [500, 1500, 2500, 800, 1200]

1. Use map() + lambda to add 18% GST

In [32]:
prices = [500, 1500, 2500, 800, 1200]

gst_prices = list(map(lambda p: p * 1.18, prices))

print(gst_prices)

[590.0, 1770.0, 2950.0, 944.0, 1416.0]


2. Use filter() to select prices above 1000

In [33]:
prices = [500, 1500, 2500, 800, 1200]
high_prices = list(filter(lambda p: p > 1000, prices))

print(high_prices)

[1500, 2500, 1200]


Question 9:

Write a generator that yields squares of numbers from 1 to 1,000,000.

Explain difference between:

● List comprehension

● Generator expression
Why are generators preferred in big data scenarios?

Answer9:


Generator that yields squares from 1 to 1,000,000

In [34]:
def square_generator():
    for i in range(1, 1_000_001):
        yield i * i

# Example usage
gen = square_generator()

print(next(gen))  # 1
print(next(gen))  # 4
print(next(gen))  # 9

1
4
9


Difference Between List Comprehension and Generator Expression

1. List Comprehension


    Creates and stores the entire list in memory immediately.

In [ ]:
squares_list = [x**2 for x in range(1, 1_000_001)]

Characteristics:

.Faster access to elements.

.Uses more memory.

.Suitable for small to medium datasets.

.All values are generated at once.

Example:

In [ ]:
print(squares_list[0])  # 1
print(squares_list[1])  # 4

2. Generator Expression

Generates values lazily (on demand).

In [ ]:
squares_gen = (x**2 for x in range(1, 1_000_001))

Characteristics:

.Memory efficient.

.Values are generated only when needed

.Suitable for large datasets or streams.

.Can be iterated only once.

Example:

In [ ]:
print(next(squares_gen))  # 1
print(next(squares_gen))  # 4

Why Generators Are Preferred in Big Data Scenarios

Suppose you need squares of 1,000,000 numbers:

List Comprehension

In [ ]:
squares = [x**2 for x in range(1, 1_000_001)]

.Creates 1,000,000 values immediately.

.Consumes significant memory.

.Can cause memory issues with much larger datasets.

Generator

In [ ]:
squares = (x**2 for x in range(1, 1_000_001))

.Stores only the current value and state.

.Memory usage remains almost constant.

.Can process datasets larger than available RAM.


Real-World Big Data Uses

Generators are commonly used for:

    .Reading huge log files line by line.

    .Processing millions of database records.

    .Streaming sensor/IoT data.

    .ETL (Extract–Transform–Load) pipelines.

    .Machine learning data loaders.

    .Processing large CSV or JSON files.

Example:

In [39]:
def read_large_file(filename):
    with open(filename) as f:
        for line in f:
            yield line

Instead of loading a 10 GB file into memory, the generator processes one line at a time, making it highly scalable and memory efficient.

Question 10:

 Design a class:
class Product:
With:

● Attributes: name, price, quantity

● Method: calculate_revenue()

Create 3 objects and print their revenue.

Answer10:

In [40]:
class Product:
    def __init__(self, name, price, quantity):
        self.name = name
        self.price = price
        self.quantity = quantity

    def calculate_revenue(self):
        return self.price * self.quantity


# Creating 3 objects
product1 = Product("Laptop", 50000, 5)
product2 = Product("Mobile", 20000, 10)
product3 = Product("Headphones", 1500, 20)

# Printing revenue
print(product1.name, "Revenue =", product1.calculate_revenue())
print(product2.name, "Revenue =", product2.calculate_revenue())
print(product3.name, "Revenue =", product3.calculate_revenue())

Laptop Revenue = 250000
Mobile Revenue = 200000
Headphones Revenue = 30000


The calculate_revenue() method calculates revenue using:

Revenue=Price×Quantity